custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go

```diff
@@ -265,6 +358,87 @@ func (t *Translator) legacyFilterForPods(podIDs []string) string {
+		case selection.Equals, selection.DoubleEquals:
+			if isAllowedLabelName(req.Key()) {
+				filters = append(filters, fmt.Sprintf("%s = %q", req.Key(), req.Values().List()[0]))
+			} else {
+				return "", provider.NewLabelNotAllowedError(req.Key())
+			}
+		case selection.NotEquals:
+			if isAllowedLabelName(req.Key()) {
+				filters = append(filters, fmt.Sprintf("%s != %q", req.Key(), req.Values().List()[0]))
+			} else {
+				return "", provider.NewLabelNotAllowedError(req.Key())
+			}
+		case selection.In:
+			if isAllowedLabelName(req.Key()) {
+				filters = append(filters, fmt.Sprintf("%s = one_of(%s)", req.Key(), strings.Join(quoteAll(req.Values().List()), ",")))
+			} else {
+				return "", provider.NewLabelNotAllowedError(req.Key())
+			}
+		case selection.NotIn:
+			if isAllowedLabelName(req.Key()) {
+				filters = append(filters, fmt.Sprintf("NOT %s = one_of(%s)", req.Key(), strings.Join(quoteAll(req.Values().List()), ",")))
+			} else {
+				return "", provider.NewLabelNotAllowedError(req.Key())
+			}
+		case selection.Exists:
+			prefix, suffix, err := splitMetricLabel(req.Key())
+			if err == nil {
+				filters = append(filters, fmt.Sprintf("%s : %s", prefix, suffix))
+			} else {
+				return "", provider.NewLabelNotAllowedError(req.Key())
+			}
+		case selection.DoesNotExist:
+			// DoesNotExist is not allowed due to Stackdriver filtering syntax limitation
+			return "", apierr.NewBadRequest("Label selector with operator DoesNotExist is not allowed")
+		case selection.GreaterThan:
+			if isAllowedLabelName(req.Key()) {
+				value, err := strconv.ParseInt(req.Values().List()[0], 10, 64)
+				if err != nil {
+					return "", apierr.NewInternalError(fmt.Errorf("Unexpected error: value %s could not be parsed to integer", req.Values().List()[0]))
+				}
+				filters = append(filters, fmt.Sprintf("%s > %v", req.Key(), value))
+			} else {
+				return "", provider.NewLabelNotAllowedError(req.Key())
+			}
+		case selection.LessThan:
+			if isAllowedLabelName(req.Key()) {
+				value, err := strconv.ParseInt(req.Values().List()[0], 10, 64)
+				if err != nil {
+					return "", apierr.NewInternalError(fmt.Errorf("Unexpected error: value %s could not be parsed to integer", req.Values().List()[0]))
+				}
+				filters = append(filters, fmt.Sprintf("%s < %v", req.Key(), value))
+			} else {
+				return "", provider.NewLabelNotAllowedError(req.Key())
+			}
+		default:
+			return "", provider.NewOperationNotSupportedError(fmt.Sprintf("Selector with operator %q", req.Operator()))
+		}
+	}
+	return strings.Join(filters, " AND "), nil
+}
+
+func (t *Translator) getMetricLabels(series *stackdriver.TimeSeries) map[string]string {
+	metricLabels := map[string]string{}
+	for label, value := range series.Metric.Labels {
+		metricLabels["metric.label."+label] = value
+	}
+	metricLabels["resource.type"] = series.Resource.Type
+	for label, value := range series.Resource.Labels {
+		metricLabels["resource.label."+label] = value
+	}
+	return metricLabels
```

Comment:
Should we include metadata labels here? I don't know if we should, but we do allow them in query.

Reply:
I think we should not. Metric and resource labels are exported with each metric. Metadata labels are exported by different channel to provide more information about given resource to Stackdriver. But resource labels include a complete set of information required to identify the resource.

For example, pods are identified by resource labels: cluster name, location, project id, namespace and name. Metadata contain information such as node name, pod labels, etc.

Reply:
So you can use metadata labels to query based on pod labels, and every metric value will come with a set of labels identifying it (namespace + name), correct? Makes sense.